esp/esdllib/esdl_xpath_extensions_libxml.cpp

```diff
@@ -472,6 +474,326 @@ static void strTokenizeFunction(xmlXPathParserContextPtr ctxt, int nargs)
+    xmlFree(toDecrypt);
+}
+
+static void encodeBase64String(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (0 == nargs || nargs > 2)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toEncode = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+    bool lineBreaks = false;
+    if (2 == nargs)
+    {
+        lineBreaks = xmlXPathPopBoolean(ctxt);
+        if (xmlXPathCheckError(ctxt))
+        {
+            xmlFree(toEncode);
+            return;
+        }
+    }
+
+    StringBuffer encoded;
+    JBASE64_Encode((const char*)toEncode, long(strlen((const char*)toEncode)), encoded, lineBreaks);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)encoded.str()));
+    xmlFree(toEncode);
+}
+
+static void decodeBase64String(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toDecode = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer decoded;
+    JBASE64_Decode((const char*)toDecode, decoded);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)decoded.str()));
+    xmlFree(toDecode);
+}
+
+static void escapeXmlCharacters(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toEncode = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer encoded;
+    encodeXML((const char*)toEncode, encoded);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)encoded.str()));
+    xmlFree(toEncode);
+}
+
+static void unescapeXmlCharacters(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toDecode = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer decoded;
+    decodeXML((const char*)toDecode, decoded);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)decoded.str()));
+    xmlFree(toDecode);
+}
+
+static void compressString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toCompress = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt))
+        return;
+
+    size32_t len = size32_t(strlen((const char *)toCompress)) + 1;
+    MemoryBuffer compressed;
+    compressToBuffer(compressed, len, (void *)toCompress);
+    StringBuffer encoded;
+    JBASE64_Encode((void*)compressed.bufferBase(), compressed.length(), encoded, false);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)encoded.str()));
+
+    xmlFree(toCompress);
+}
+
+static void decompressString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
```

Comment:
Have you tested what happens if either the base64 or compressed data isn't valid?

Reply:
I had not. I'm not concerned about bad encoding because even a good encoding can yield invalid compressed data. I'll catch exceptions from the decompress.

Reply:
Did you mean you are going to add a try to this PR?

Reply:
Yes. The decompress call does throw. When I passed "this is bad data" to the XPath function the decode "completed" and the decompress threw an exception.